### PR TITLE
curvefs/metaserver: fix trash bugs

### PR DESCRIPTION
**Pull request by wuhongsong**
### What problem does this PR solve?

Issue Number: #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by wuhongsong** [comment]
cicheck

File: curvefs/src/metaserver/inode_storage.h [hunk 1]
```diff
@@ -76,6 +76,16 @@ class InodeStorage {
      */
     MetaStatusCode Insert(const Inode& inode, int64_t logIndex);
 
+    /**
+     * @brief update deleting inode key in storage
+     * @param[in] inode: the inode want to update
+     * @param[in] logIndex: the index of raft log
+     * @return
+     */
+    MetaStatusCode UpdateDeletingKey(const Inode& inode, int64_t logIndex);
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_storage.h:85]: `UpdateDeletingKey` to `AddDeletedInode` maybe more clearer

File: curvefs/src/metaserver/inode_storage.h [hunk 2]
```diff
@@ -76,6 +76,16 @@ class InodeStorage {
...
+
+    MetaStatusCode ClearDelKey(const Key4Inode& key);
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_storage.h:87]: Ditto `RemoveDeletedInode`

File: curvefs/src/metaserver/metastore.cpp [hunk 3]
```diff
@@ -866,5 +869,27 @@ bool MetaStoreImpl::InitStorage() {
     return kvStorage_->Open();
 }
 
+void MetaStoreImpl::LoadDeletedInodes() {
```
Comment (SeanHai) [inline curvefs/src/metaserver/metastore.cpp:872]: `LoadDeletedInodes` is the responsibility of InodeManager, Each partition has one InodeManager and InodeStorage, you can load deleted inodes with InodeManager and add to trash_ in InodeManager by pass and the deleted inode loaded are all belong to this partition and trash.

File: curvefs/src/metaserver/inode_storage.cpp [hunk 4]
```diff
@@ -185,9 +186,74 @@ MetaStatusCode InodeStorage::Insert(const Inode& inode, int64_t logIndex) {
...
+    std::string skey = conv_.SerializeToString(key);
+    VLOG(9) << "update deleting key, " << inode.inodeid();
+    const char* step = "Begin transaction";
+    std::shared_ptr<storage::StorageTransaction> txn;
+    txn = kvStorage_->BeginTransaction();
+    if (txn == nullptr) {
+        LOG(ERROR) << "Begin transaction failed";
+        return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+    }
+    auto rc = txn->HSet(table4DelInode_, skey , inode);
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_storage.cpp:202]: The value does not need to be the entire inode, dtime is enough

File: curvefs/src/metaserver/trash.cpp [hunk 5]
```diff
@@ -56,75 +56,99 @@ void TrashImpl::Init(const TrashOption &option) {
...
 void TrashImpl::StopScan() {
     isStop_ = true;
 }
 
-bool TrashImpl::IsStop() {
-    return isStop_;
-}
-
-void TrashImpl::Add(uint64_t inodeId, uint64_t dtime) {
+void TrashImpl::Add(uint64_t inodeId, uint64_t dtime, bool deleted) {
```
Comment (SeanHai) [inline curvefs/src/metaserver/trash.cpp:67]: Why `deleted` needed?

File: curvefs/src/metaserver/trash.cpp [hunk 6]
```diff
@@ -56,75 +56,99 @@ void TrashImpl::Init(const TrashOption &option) {
...
 void TrashImpl::Remove(uint64_t inodeId) {
     if (isStop_) {
         return;
     }
+    TrashItem item;
+    item.inodeId = inodeId;
     LockGuard lg(itemsMutex_);
-    trashItems_.erase(inodeId);
+    auto it = std::find(trashItems_.begin(), trashItems_.end(), item);
+    if (it != trashItems_.end()) {
```
Comment (SeanHai) [inline curvefs/src/metaserver/trash.cpp:95]: The remove of rocksdb data can do here also.

File: curvefs/src/metaserver/inode_manager.cpp [hunk 7]
```diff
@@ -446,6 +441,13 @@ MetaStatusCode InodeManager::UpdateInode(const UpdateInodeRequest& request,
             return MetaStatusCode::STORAGE_INTERNAL_ERROR;
         }
     }
+
+    if (needAddTrash) {
+        inodeStorage_->UpdateDeletingKey(old, logIndex);
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_manager.cpp:446]: add and delete op all deal in trash is better.

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

File: curvefs/src/metaserver/inode_storage.h (hunk 1 above)
Comment (wuhongsong) [reply curvefs/src/metaserver/inode_storage.h]: fix

File: curvefs/src/metaserver/inode_storage.h (hunk 2 above)
Comment (wuhongsong) [reply curvefs/src/metaserver/inode_storage.h]: fix

File: curvefs/src/metaserver/metastore.cpp (hunk 3 above)
Comment (wuhongsong) [reply curvefs/src/metaserver/metastore.cpp:872]: fix

File: curvefs/src/metaserver/trash.cpp (hunk 6 above)
Comment (wuhongsong) [reply curvefs/src/metaserver/trash.cpp:98]: fix

File: curvefs/src/metaserver/inode_manager.cpp (hunk 7 above)
Comment (wuhongsong) [reply curvefs/src/metaserver/inode_manager.cpp]: fix

File: curvefs/src/metaserver/inode_storage.cpp (hunk 4 above)
Comment (wuhongsong) [reply curvefs/src/metaserver/inode_storage.cpp]: fix

File: curvefs/src/metaserver/trash.cpp (hunk 5 above)
Comment (wuhongsong) [reply curvefs/src/metaserver/trash.cpp:67]: tell reboot or not

**Review by xu-chaojie — APPROVED** [review]

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

File: curvefs/src/metaserver/inode_manager.cpp [hunk 8]
```diff
@@ -446,10 +441,27 @@ MetaStatusCode InodeManager::UpdateInode(const UpdateInodeRequest& request,
...
 }
 
+void InodeManager::LoadDeletedInodes() {
+    std::list<std::string> items;
+    inodeStorage_->LoadDeletedInodes(&items);
+    VLOG(3) << "build trash items size: " << items.size();
+    std::vector<std::string> names;
+    for (auto iter : items) {
+        curve::common::SplitString(iter , ":", &names);
+        trash_->Add(std::stoull(names[names.size() - 1 ]), 0, true);
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_manager.cpp:461]: Why set dtime to 0? The dtime lost when you add to rocksdb.

File: curvefs/src/metaserver/trash.h [hunk 9]
```diff
@@ -52,6 +52,19 @@ using ::curve::common::InterruptibleSleeper;
 using ::curvefs::client::rpcclient::MdsClient;
 using ::curvefs::client::rpcclient::MdsClientImpl;
 
+struct TrashItem {
```
Comment (SeanHai) [inline curvefs/src/metaserver/trash.h:55]: The `TrashItem` is not need. Map<inodeId, dtime> is enough and better to find process when delete.

File: curvefs/src/metaserver/inode_storage.cpp [hunk 10]
```diff
@@ -185,9 +186,74 @@ MetaStatusCode InodeStorage::Insert(const Inode& inode, int64_t logIndex) {
...
+    if (!txn->Rollback().ok()) {
+        LOG(ERROR) << "Rollback delete inode transaction failed, status = "
+                   << s.ToString();
+    }
+    return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+}
+
+void InodeStorage::LoadDeletedInodes(std::list<std::string> * inodes) {
+    VLOG(6) << "load deleted key start with: " << table4DelInode_;
+    kvStorage_->GetPrefix(inodes, table4DelInode_);
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_storage.cpp:249]: The parse of key and value can do here and return the final result map<inodeId, dtime> is ok.

File: curvefs/test/metaserver/trash_test.cpp [hunk 11]
```diff
@@ -215,9 +215,9 @@ TEST_F(TestTrash, testAdd3ItemAndDelete) {
         }));
 
     uint64_t dtime = curve::common::TimeUtility::GetTimeofDaySec();
-    trash1->Add(1, dtime - 6);
-    trash1->Add(2, dtime - 2);
-    trash2->Add(1, dtime);
+    trash1->Add(1, dtime - 6, false);
+    trash1->Add(2, dtime - 2, false);
+    trash2->Add(1, dtime, false);
```
Comment (SeanHai) [inline curvefs/test/metaserver/trash_test.cpp:220]: It is best to test the load deleted inode interface.

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

File: curvefs/src/metaserver/storage/converter.h [hunk 12]
```diff
@@ -34,9 +34,10 @@ namespace curvefs {
 namespace metaserver {
 
 class MetaStoreFStream;
-
 namespace storage {
 
+#define DELETEPREFIXINKV "deleting:"
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/converter.h:39]: this is needn't

File: curvefs/src/metaserver/storage/converter.cpp [hunk 13]
```diff
@@ -123,6 +129,13 @@ std::string NameGenerator::Format(KEY_TYPE type, uint32_t partitionId) {
     return absl::StrCat(type, kDelimiter, absl::string_view(buf, sizeof(buf)));
 }
 
+std::string NameGenerator::FormatDel(KEY_TYPE type, uint32_t partitionId) {
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/converter.cpp:152]: don't need

File: curvefs/src/metaserver/storage/converter.cpp [hunk 14]
```diff
@@ -54,6 +55,7 @@ static bool CompareType(const std::string& str, KEY_TYPE keyType) {
 
 NameGenerator::NameGenerator(uint32_t partitionId)
     : tableName4Inode_(Format(kTypeInode, partitionId)),
+      tableName4DelInode_(FormatDel(kTypeInode, partitionId)),
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/converter.cpp:61]: Add a new type like 'kTypeDelInode'

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp [hunk 15]
```diff
@@ -547,6 +553,31 @@ bool RocksDBStorage::Recover(const std::string& dir) {
...
+void  RocksDBStorage::GetPrefix(
+  std::map<std::string, uint64_t>* item, const std::string prefix) {
+    std::string sprefix = absl::StrCat("0", ":", prefix);
+    VLOG(3) << "load deleted inodes from: " << options_.dataDir
+      << ", " << sprefix << ", " << prefix;
+    int counts = 0;
+    rocksdb::Iterator* it = db_->NewIterator(rocksdb::ReadOptions());
+    curvefs::metaserver::Time time;
+    for (it->Seek(sprefix); it->Valid() &&
+      it->key().starts_with(sprefix); it->Next()) {
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/rocksdb_storage.cpp:565]: Ditto

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp [hunk 16]
```diff
@@ -547,6 +553,31 @@ bool RocksDBStorage::Recover(const std::string& dir) {
...
+        std::string key = it->key().ToString();
+       if (!time.ParseFromString(it->value().ToString())) {
+            return;
+       }
+       VLOG(9) << "key: " << key << ", " << key.size() << ", "
+           << it->value().ToString() << ", " << time.sec();
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/rocksdb_storage.cpp:571]: Ditto

File: curvefs/src/metaserver/storage/rocksdb_storage.cpp [hunk 17]
```diff
@@ -547,6 +553,31 @@ bool RocksDBStorage::Recover(const std::string& dir) {
...
+        item->emplace(key, time.sec());
+        counts++;
+    }
+    delete it;
+    VLOG(3) << "load deleted inodes end, size is: " << item->size()
+    << ", " << counts << ", " << options_.dataDir;
```
Comment (SeanHai) [inline curvefs/src/metaserver/storage/rocksdb_storage.cpp:577]: Ditto

File: curvefs/src/metaserver/inode_storage.cpp [hunk 18]
```diff
@@ -185,9 +187,78 @@ MetaStatusCode InodeStorage::Insert(const Inode& inode, int64_t logIndex) {
...
+        LOG(ERROR) << "Rollback delete inode transaction failed, status = "
+                   << s.ToString();
+    }
+    return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+}
+
+void InodeStorage::LoadDeletedInodes(std::map<std::string, uint64_t> * inodes) {
+    VLOG(6) << "load deleted key start with: " << table4DelInode_;
+    kvStorage_->GetPrefix(inodes, table4DelInode_);
+    VLOG(6) << "oad deleted over";
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_storage.cpp:255]: load

File: curvefs/src/metaserver/inode_manager.cpp [hunk 19]
```diff
@@ -446,10 +441,27 @@ MetaStatusCode InodeManager::UpdateInode(const UpdateInodeRequest& request,
...
     VLOG(9) << "UpdateInode success, " << request.ShortDebugString();
     return MetaStatusCode::OK;
 }
 
+void InodeManager::LoadDeletedInodes() {
+    std::map<std::string, uint64_t> items;
+    inodeStorage_->LoadDeletedInodes(&items);
+    VLOG(3) << "build trash items size: " << items.size();
+    std::vector<std::string> names;
+    for (auto iter : items) {
```
Comment (SeanHai) [inline curvefs/src/metaserver/inode_manager.cpp:459]: `for (auto& iter : items)`

File: curvefs/src/metaserver/trash.cpp [hunk 20]
```diff
@@ -116,24 +125,30 @@ void TrashImpl::ScanTrash() {
                 LOG(ERROR) << "DeleteInodeAndData fail, fsId = " << fsId_
                            << ", inodeId = " << it->first
                            << ", ret = " << MetaStatusCode_Name(ret);
+                RemoveDeletedInode(it->first);
```
Comment (SeanHai) [inline curvefs/src/metaserver/trash.cpp:128]: RemoveDeleteInode in func `Remove`

File: curvefs/src/metaserver/trash.cpp [hunk 21]
```diff
@@ -116,24 +125,30 @@ void TrashImpl::ScanTrash() {
...
     }
-
     {
         LockGuard lgItems(itemsMutex_);
-        trashItems_.insert(temp.begin(), temp.end());
+        trashItems_.swap(temp);
     }
 }
 
+void TrashImpl::RemoveDeletedInode(uint64_t inodeId) {
```
Comment (SeanHai) [inline curvefs/src/metaserver/trash.cpp:147]: Maybe void is not good. Some warning log may need if Remove fail.

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by YunhuiChen** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Review by SeanHai — APPROVED** [review]

**Pull request merged**